packages/firestore/lite/src/api/reference.ts

```diff
@@ -300,6 +306,40 @@ export function getDoc<T>(
   });
 }
 
+export function setDoc<T>(
+  reference: firestore.DocumentReference<T>,
+  data: T,
+  options?: firestore.SetOptions
+): Promise<void> {
+  const ref = tryCast(reference, DocumentReference);
+
+  const [convertedValue] = applyFirestoreDataConverter(
+    ref._converter,
+    data,
+    'setDoc'
+  );
+
+  // Kick off configuring the client, which freezes the settings.
+  const configureClient = ref.firestore._ensureClientConfigured();
+  const dataReader = newUserDataReader(
+    ref.firestore._databaseId,
+    ref.firestore._settings!
+  );
+
+  const parsed = isMerge(options)
+    ? dataReader.parseMergeData('setDoc', convertedValue)
+    : isMergeFields(options)
+    ? dataReader.parseMergeData('setDoc', convertedValue, options.mergeFields)
+    : dataReader.parseSetData('setDoc', convertedValue);
```

Comment:
Optional: I like the helper methods you made to interrogate the options but the ternary required effort to parse. Consider factoring this out into a helper function that includes the `toMutations()` call or perhaps convert this to a series of if/then/else.

Reply:
I have been bothered by this code for a while, as it is repeated in a couple of places in the SDK. I changed UserDataReader.parseSetData() to perform this check on behalf of the caller.